### PR TITLE
feature: new rule package-directory-mismatch

### PR DESCRIPTION
**Pull request by vbvictor**
Added new rule `package-directory-mismatch` that will find mismatch between directory name and `package` name.

Closes https://github.com/mgechev/revive/issues/1428.

### CODE REVIEW

File: rule/package_directory_mismatch.go [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+
+	packageName := file.AST.Name.Name
+	if packageName == "main" {
+		return nil
+	}
+
+	dirPath := filepath.Dir(file.Name)
+	dirName := filepath.Base(dirPath)
+
+	if dirName == "." || dirName == "/" || dirName == "" {
```
Comment (chavacava) [inline rule/package_directory_mismatch.go]: We could create a map with all the dirs names to skip (. / test data...) and use the map to check

Comment (vbvictor) [reply rule/package_directory_mismatch.go]: Done

**Review by chavacava — COMMENTED** [review]
Hi @vbvictor (thanks for the PR!)
Could you please update the readme and rules descriptions documents ?

(right now I have limited internet access thus let's wait for feedback from @alexandear  @denisvmedia @cc)

File: rule/package_directory_mismatch.go [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+// Apply applies the rule to the given file.
+func (*PackageDirectoryMismatchRule) Apply(file *lint.File, _ lint.Arguments) []lint.Failure {
+	if file.IsTest() {
+		return nil
+	}
+
+	packageName := file.AST.Name.Name
+	if packageName == "main" {
+		return nil
+	}
```
Comment (ccoVeille) [inline rule/package_directory_mismatch.go]: I would use IsMain method here

File: rule/package_directory_mismatch.go [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+			{
+				Failure:    fmt.Sprintf("package name %q does not match directory name %q", packageName, dirName),
+				Confidence: 1,
+				Node:       file.AST.Name,
+				Category:   lint.FailureCategoryNaming,
+			},
+		}
+	}
+
+	return nil
```
Comment (ccoVeille) [inline rule/package_directory_mismatch.go]: Please consider inverting the logic

```go
if packageName == dirName {
    return nil
}

return []lint.Failure{..........
```

File: rule/package_directory_mismatch.go (hunk 2 above)
Comment (vbvictor) [reply rule/package_directory_mismatch.go]: Added
```go
if file.IsTest() || file.Pkg.IsMain() {
  return nil
}
```
`if` is fairly simple, so I put two checks in one.

File: rule/package_directory_mismatch.go [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+func (*PackageDirectoryMismatchRule) Apply(file *lint.File, _ lint.Arguments) []lint.Failure {
+	if file.IsTest() || file.Pkg.IsMain() {
+		return nil
+	}
+
+	dirPath := filepath.Dir(file.Name)
+	dirName := filepath.Base(dirPath)
+	if _, found := skipDirs[dirName]; found {
+		return nil
+	}
```
Comment (ccoVeille) [inline rule/package_directory_mismatch.go]: I'm unsure this would work for a file like

testdata/noregression/foo.go

```go
package foo
```

I mean this should be ignored 👆 
It's common that testdata are full of random file, especially in Go linter code.

Please note, that on the contrary this should be detected 👇

internal/foo/something.go

```go
package whatever // MATCH
```

And this shouldn't 👇

internal/bar/quz.go
```go
package bar
```

Comment (ccoVeille) [reply rule/package_directory_mismatch.go]: Please note, I'm not even sure whether this is valid, nor if it should be reported

internal/abcd.go
```go
package bar
```

I have no computer with me while I'm doing the review. Please check if this file can be imported and what this package should be

Any idea @alexandear ?

**Review by alexandear — CHANGES_REQUESTED** [review]
Let's handle also directories like `v2, v3 ...` etc.

What I mean:

| **directory structure** | **package name** | **reference** |
|-------------------------|------------------|---------------|
| `math/rand/v2`| `rand` | https://go.googlesource.com/go/+/refs/tags/go1.24.0/src/math/rand/v2 
| `encoding/json/v2` | `json` | https://go.googlesource.com/go/+/refs/tags/go1.25rc2/src/encoding/json/v2/ |

File: rule/package_directory_mismatch.go (hunk 4 above)
Comment (alexandear) [reply rule/package_directory_mismatch.go]: > Please note, I'm not even sure whether this is valid, nor if it should be reported
> 
> internal/abcd.go
> 
> ```go
> package bar
> ```
> 
> I have no computer with me while I'm doing the review. Please check if this file can be imported and what this package should be
> 
> Any idea @alexandear ?

This is valid, and I think shouldn't be reported. In the future, we can extend the rule to detect such cases.

Comment (alexandear) [reply rule/package_directory_mismatch.go]: > I'm unsure this would work for a file like
> 
> testdata/noregression/foo.go
> 
> ```go
> package foo
> ```
> 
> I mean this should be ignored 👆 It's common that testdata are full of random file, especially in Go linter code.

Agree. Let's ignore any matches inside `testdata`.

**Comment by alexandear** [comment]
To reduce noise, maybe it makes sense to first exclude `-`, `_`, and `.` from a directory name and then compare with the package name. The following should not be reported:

| **directory** | **package name** | **reference** |
|-------------------------|------------------|---------------|
| `remote-exec`| `remoteexec` | https://github.com/hashicorp/terraform/blob/d1e412fcf0eb29be0fe9f433f756466067e424f4/internal/builtin/provisioners/remote-exec/resource_provisioner.go#L4 |
| `e2e_node` | `e2enode` | https://github.com/kubernetes/kubernetes/blob/963a9acb6d88cc37733140d140bffa8966f4d90d/test/e2e_node/image_volume.go#L17 |
| `cilium.io` | `ciliumio` | https://github.com/cilium/cilium/blob/dd2295f795974ee05dbb84b451a9a7635c5bc990/pkg/k8s/apis/cilium.io/register.go#L4 |

**Comment by vbvictor** [comment]
I created option "IgnoredRegex" to bypass our `testdata/` directory that ignored all my tests.

File: rule/package_directory_mismatch.go [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+
+	return nil
+}
+
+// normalizeName removes hyphens, underscores, and dots from the name
+// to allow matching between directory names like "foo-bar.buz" and package names like "foobarbuz".
+func normalizeName(name string) string {
+	name = strings.ReplaceAll(name, "-", "")
+	name = strings.ReplaceAll(name, "_", "")
+	name = strings.ReplaceAll(name, ".", "")
```
Comment (ccoVeille) [inline rule/package_directory_mismatch.go]: In such a case I tend to use a strings.Replacer variable at the package level.

https://pkg.go.dev/strings#NewReplacer

I think it brings clarity, but it's opinionated somehow

File: rule/package_directory_mismatch.go [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+
+// skipDirs contains directory names that should be unconditionally ignored when checking.
+var skipDirs = map[string]struct{}{
+	".": {},
+	"/": {},
+	"":  {},
+}
+
+// isVersionDirectory checks if a directory name is a version directory (v1, v2, etc.)
+func isVersionDirectory(name string) bool {
```
Comment (ccoVeille) [inline rule/package_directory_mismatch.go]: It's more a path than a directory.

But I'm nitpicking 🤣

File: rule/package_directory_mismatch.go [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+	dirPath := filepath.Dir(file.Name)
+	dirName := filepath.Base(dirPath)
+
+	if r.ignoredDirs != nil && r.ignoredDirs.MatchString(dirPath) {
+		return nil
+	}
+
+	// Use the parent directory for comparison if the immediate directory is a version directory (v1, v2, etc.).
+	if isVersionDirectory(dirName) {
+		dirName = filepath.Base(filepath.Dir(dirPath))
```
Comment (ccoVeille) [inline rule/package_directory_mismatch.go]: Maybe a comment about the fact we use the parent path then

File: rule/package_directory_mismatch.go (hunk 5 above)
Comment (vbvictor) [reply rule/package_directory_mismatch.go]: Thank you, didn't know about such class.

File: rule/package_directory_mismatch.go (hunk 7 above)
Comment (vbvictor) [reply rule/package_directory_mismatch.go]: Placed comment right above `dirName = `

**Review by ccoVeille — APPROVED** [review]

**Comment by vbvictor** [comment]
As noted in https://github.com/mgechev/revive/pull/1445#issuecomment-3124604709, when we have https://pkg.go.dev/k8s.io/api/core/v1 we will get false-positive because package name is actually `v1` and directory will be `core`. When dealing with version packages, I think we should check for `packageName == dirName || packageName == prevDirName`.

**Comment by ccoVeille** [comment]
> As noted in https://github.com/mgechev/revive/pull/1445#issuecomment-3124604709, when we have https://pkg.go.dev/k8s.io/api/core/v1 we will get false-positive because package name is actually `v1` and directory will be `core`. When dealing with version packages, I think we should check for `packageName == dirName || packageName == prevDirName`.

Here we are in an edge case, but if there is a way to support it, yes, why not

File: rule/package_directory_mismatch.go [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+	}
+
+	for i := 1; i < len(name); i++ {
+		if name[i] < '0' || name[i] > '9' {
+			return false
+		}
+	}
+
+	return true
+}
```
Comment (ccoVeille) [inline rule/package_directory_mismatch.go]: Because this is used in #1429 and #1445

It should be moved and used by both code. There is no need to duplicate code, tests, with ygr risk of maintaining them separately

Comment (vbvictor) [reply rule/package_directory_mismatch.go]: I think I Can make a small NFC PR with this function and rebase both PRs after merge

Comment (vbvictor) [reply rule/package_directory_mismatch.go]: Otherwise, just wait until one of these two PRs are merged and the next one will refactor this function

File: RULES_DESCRIPTIONS.md [hunk 9]
```diff
@@ -921,6 +922,50 @@ More [information here](https://go.dev/wiki/CodeReviewComments#package-comments)
...
+```toml
+[rule.package-directory-mismatch]
+arguments = ["testdata"]
+```
+
+Exclude specific directories
+
+```toml
+[rule.package-directory-mismatch]
+arguments = ["(testcases|testinfo)"]
```
Comment (alexandear) [inline RULES_DESCRIPTIONS.md]: Based on my experience maintaining Revive rules, it's better to name the options and use slices for the values:

```suggestion
arguments = [{ ignore-directories = ["testcases", "testinfo"] }]
```

File: RULES_DESCRIPTIONS.md [hunk 10]
```diff
@@ -921,6 +922,50 @@ More [information here](https://go.dev/wiki/CodeReviewComments#package-comments)
...
+
+_Configuration_: (string) regular expression for filepaths to be excluded. Default: `"testdata"`.
+
+Examples:
+
+Default behavior (excludes paths containing "testdata")
+
+```toml
+[rule.package-directory-mismatch]
+arguments = ["testdata"]
```
Comment (alexandear) [inline RULES_DESCRIPTIONS.md]: Default == no arguments:

```suggestion
Default behavior (excludes paths containing "testdata")

```toml
[rule.package-directory-mismatch]
```

Please rearrange examples a little bit.

File: RULES_DESCRIPTIONS.md (hunk 9 above)
Comment (alexandear) [reply RULES_DESCRIPTIONS.md]: @chavacava what do you think?

Comment (chavacava) [reply RULES_DESCRIPTIONS.md]: > Based on my experience maintaining Revive rules, it's better to name the options and use slices for the values

I agree 100%

Comment (chavacava) [reply RULES_DESCRIPTIONS.md]: Side note on ignoring `testdata`: IMO `revive` should follow standard go tools and ignore everything under `testdata`

See [here](https://pkg.go.dev/cmd/go#hdr-Test_packages):
> The go tool will ignore a directory named "testdata", making it available to hold ancillary data needed by the tests.

Comment (vbvictor) [reply RULES_DESCRIPTIONS.md]: > Side note on ignoring `testdata`: IMO `revive` should follow standard go tools and ignore everything under `testdata`

The only problem is that `revive` tests are in `testdata` so `package-directory-mismatch` (and all other rules) will ignore their tests. I had to make this configurable to be able to run tests.

File: RULES_DESCRIPTIONS.md [hunk 11]
```diff
@@ -921,6 +922,50 @@ More [information here](https://go.dev/wiki/CodeReviewComments#package-comments)
...
 
+## package-directory-mismatch
+
+_Description_: It is considered a good practice to name a package after the directory containing the file.
+This rule warns when the package name declared in the file does not match the name of the directory containing the file.
+
+The following cases are excluded from this check:
+
+- Package `main` (executable packages)
+- Test files (files with `_test` suffix)
```
Comment (denisvmedia) [inline RULES_DESCRIPTIONS.md]: Why to exclude them?

File: RULES_DESCRIPTIONS.md [hunk 12]
```diff
@@ -921,6 +922,50 @@ More [information here](https://go.dev/wiki/CodeReviewComments#package-comments)
...
+- Files in `testdata` directories (at any level) - by default
+- Files directly in `internal` directories (but files in subdirectories of `internal` are checked)
+- Files in version directories (`v1`, `v2`, etc.) - the parent directory name is used for comparison
```
Comment (denisvmedia) [inline RULES_DESCRIPTIONS.md]: There are packages with literally v1/v2 names (also e.g. v1alpha1). I'd first attempt to match them only if no match, do what you propose here.

File: RULES_DESCRIPTIONS.md [hunk 13]
```diff
@@ -921,6 +922,50 @@ More [information here](https://go.dev/wiki/CodeReviewComments#package-comments)
...
+
+The rule normalizes directory and package names before comparison by removing hyphens (`-`), underscores (`_`), and dots (`.`). This allows packages like:
+
+- Directory `foo-bar` with package `foobar`
+- Directory `foo_bar` with package `foobar`
```
Comment (denisvmedia) [inline RULES_DESCRIPTIONS.md]: Although not recommended, it is possible to have a ppackage named `foo_bar`. I think this rule should first check for it. Linting the package name on `_` should be done elsewhere.

Comment (vbvictor) [reply RULES_DESCRIPTIONS.md]: In the check, I "normalize" both paths: package and directory, so the following are equal:
`package foo_bar == dir foo-bar`
`package foobar == dir foo.bar`

Comment (vbvictor) [reply RULES_DESCRIPTIONS.md]: I'll adjust description to match this behavoir

File: RULES_DESCRIPTIONS.md (hunk 11 above)
Comment (vbvictor) [reply RULES_DESCRIPTIONS.md]: We generally can have directory `foo` and tests with `package foo_test`.
I adjust rule to check if test file have `package` equal to dir or `package - _test` equal to dir

Comment (denisvmedia) [reply RULES_DESCRIPTIONS.md]: I agree with the change.

File: rule/package_directory_mismatch.go [hunk 14]
```diff
@@ -18,63 +18,62 @@ const defaultIgnoredDirs = "testdata"
 
 // Configure validates the rule configuration, and configures the rule accordingly.
 func (r *PackageDirectoryMismatchRule) Configure(arguments lint.Arguments) error {
-	ignoredDirs := defaultIgnoredDirs
+	r.ignoredDirs = nil
```
Comment (denisvmedia) [inline rule/package_directory_mismatch.go]: Why not to rely on a zero value, which is already nil?

Comment (vbvictor) [reply rule/package_directory_mismatch.go]: removed unused `nil`'s

**Comment by chavacava** [comment]
I've ran the rule on some code bases and it works fine.
One pattern that seems to be popular is to have a directory `go-foo` for the package `foo`.
For example:
```
./kubesphere/vendor/github.com/stoewer/go-strcase/doc.go:8:9: package name "strcase" does not match directory name "go-strcase"
./kubesphere/vendor/github.com/mattn/go-isatty/isatty_tcgets.go:6:9: package name "isatty" does not match directory name "go-isatty"
./kubesphere/vendor/github.com/mattn/go-colorable/colorable_others.go:4:9: package name "colorable" does not match directory name "go-colorable"
./kubesphere/vendor/github.com/docker/go-units/ulimit.go:1:9: package name "units" does not match directory name "go-units"
./kubesphere/vendor/github.com/grpc-ecosystem/go-grpc-prometheus/metric_options.go:1:9: package name "grpc_prometheus" does not match directory name "go-grpc-prometheus"
./kubesphere/vendor/github.com/go-jose/go-jose/v4/doc.go:25:9: package name "jose" does not match directory name "v4" or parent directory name "go-jose"
./kubesphere/vendor/github.com/jmespath/go-jmespath/util.go:1:9: package name "jmespath" does not match directory name "go-jmespath"
./kubesphere/vendor/github.com/hashicorp/go-multierror/flatten.go:1:9: package name "multierror" does not match directory name "go-multierror"
./kubesphere/vendor/github.com/docker/go-metrics/register.go:1:9: package name "metrics" does not match directory name "go-metrics"
./kubesphere/vendor/github.com/Azure/go-ntlmssp/version.go:1:9: package name "ntlmssp" does not match directory name "go-ntlmssp"
./kubesphere/vendor/github.com/emicklei/go-restful/v3/compressors.go:1:9: package name "restful" does not match directory name "v3" or parent directory name "go-restful"
./kubesphere/vendor/github.com/monochromegane/go-gitignore/initial_holder.go:1:9: package name "gitignore" does not match directory name "go-gitignore"
./kubesphere/vendor/github.com/opencontainers/go-digest/digest.go:16:9: package name "digest" does not match directory name "go-digest"
./kubesphere/vendor/github.com/pelletier/go-toml/v2/decode.go:1:9: package name "toml" does not match directory name "v2" or parent directory name "go-toml"
./kubesphere/vendor/github.com/mattn/go-runewidth/runewidth_posix.go:4:9: package name "runewidth" does not match directory name "go-runewidth"
./kubesphere/vendor/github.com/go-git/go-git/v5/options.go:1:9: package name "git" does not match directory name "v5" or parent directory name "go-git"
./kubesphere/vendor/github.com/rcrowley/go-metrics/syslog.go:3:9: package name "metrics" does not match directory name "go-metrics"
./kubesphere/vendor/github.com/speps/go-hashids/hashids.go:7:9: package name "hashids" does not match directory name "go-hashids"
```

Those cases are about ~50% of failures on Kubesphere and ~33% on Kubernetes.
Should the rule avoid generating a failure when the package is _foo_ and the directory is _go-foo_?

**Comment by vbvictor** [comment]
> Should the rule avoid generating a failure when the package is foo and the directory is go-foo?

This seems easy to achieve so I'm +1 on this change.

File: rule/package_directory_mismatch.go [hunk 15]
```diff
@@ -0,0 +1,146 @@
...
+	".": {}, // Current directory
+	"/": {}, // Root directory
+	"":  {}, // Empty path
+}
+
+// semanticallyEqual checks if package and directory names are semantically equal to each other.
+func (PackageDirectoryMismatchRule) semanticallyEqual(packageName, dirName string) bool {
+	return normalizePath(dirName) == normalizePath(packageName) ||
+		normalizePath(dirName) == normalizePath("go"+packageName)
+}
```
Comment (chavacava) [inline rule/package_directory_mismatch.go]:
```suggestion
func (PackageDirectoryMismatchRule) semanticallyEqual(packageName, dirName string) bool {
	normDir := normalizePath(dirName)
	normPkg := normalizePath(packageName)
	return normDir == normPkg || normDir == "go"+normPkg 
}
```

File: rule/package_directory_mismatch.go [hunk 16]
```diff
@@ -0,0 +1,146 @@
...
+var skipDirs = map[string]struct{}{
+	".": {}, // Current directory
+	"/": {}, // Root directory
+	"":  {}, // Empty path
+}
+
+// semanticallyEqual checks if package and directory names are semantically equal to each other.
+func (PackageDirectoryMismatchRule) semanticallyEqual(packageName, dirName string) bool {
+	return normalizePath(dirName) == normalizePath(packageName) ||
+		normalizePath(dirName) == normalizePath("go"+packageName)
```
Comment (ccoVeille) [inline rule/package_directory_mismatch.go]: Shouldn't be this? 🤔


```suggestion
		normalizePath(dirName) == normalizePath("go-"+packageName)
```

File: rule/package_directory_mismatch.go [hunk 17]
```diff
@@ -0,0 +1,146 @@
...
+			return nil
+		}
+
+		failure = fmt.Sprintf("package name %q does not match directory name %q or parent directory name %q", packageName, dirName, parentDirName)
+	}
+
+	// When no specific message was place, use default one
+	if failure == "" {
+		failure = fmt.Sprintf("package name %q does not match directory name %q", packageName, dirName)
+	}
```
Comment (ccoVeille) [inline rule/package_directory_mismatch.go]:
```suggestion
	// define a default failure message
	failure := fmt.Sprintf("package name %q does not match directory name %q", packageName, dirName)

	// For version directories (v1, v2, etc.), we need to check also the parent directory
	if isVersionPath(dirName) {
		parentDirName := filepath.Base(filepath.Dir(dirPath))
		if r.semanticallyEqual(packageName, parentDirName) {
			return nil
		}

		failure = fmt.Sprintf("package name %q does not match directory name %q or parent directory name %q", packageName, dirName, parentDirName)
	}
```

File: rule/package_directory_mismatch.go (hunk 16 above)
Comment (chavacava) [reply rule/package_directory_mismatch.go]: Yes... and no. In fact the call to normalizePath is not necessary (unless you do "go-")

File: rule/package_directory_mismatch.go [hunk 18]
```diff
@@ -0,0 +1,146 @@
...
+		normalizePath(dirName) == normalizePath("go"+packageName)
+}
+
+// Apply applies the rule to the given file.
+func (r *PackageDirectoryMismatchRule) Apply(file *lint.File, _ lint.Arguments) []lint.Failure {
+	if file.Pkg.IsMain() {
+		return nil
+	}
+
+	dirPath := filepath.Dir(file.Name)
```
Comment (chavacava) [inline rule/package_directory_mismatch.go:99]: I think we can normalize both paths here once for all and remove further calls to `normalizePath`
As is, some cases result in calling more than once `normalizePath` on the same string.
For example, I've added some traces and we got:
```
>>>> applying to ../testdata/package_directory_mismatch/mixed/bad.go
>>>> semantically equal on: wrong mixed
>>>> normalizing mixed
>>>> normalizing wrong
>>>> normalizing mixed
>>>> normalizing gowrong
```
Notice we normalize `mixed` twice (and `gowrong` that is a string we built... thus let's build it already normalized)

File: rule/package_directory_mismatch.go (hunk 16 above)
Comment (vbvictor) [reply rule/package_directory_mismatch.go]: > In fact the call to normalizePath is not necessary (unless you do "go-")

It's still necessary to make this case work: `package foo_bar` with `dir go-foobar`. This testcase is missing, will add.

Comment (vbvictor) [reply rule/package_directory_mismatch.go]: But with this https://github.com/mgechev/revive/pull/1429#discussion_r2264628892, it's not needed anymore.

File: rule/package_directory_mismatch.go (hunk 18 above)
Comment (vbvictor) [reply rule/package_directory_mismatch.go:99]: I decided not to place normalization out of `semanticallyEqual` because having normalization inside gives convenient API for `semanticallyEqual` and make code in `Apply` more readable (e.g. we need to  normalize `parentDirName` too). Even with duplications, I think the rule will still be very fast given it doesn't traverse any AST.

File: rule/package_directory_mismatch.go [hunk 19]
```diff
@@ -0,0 +1,142 @@
...
+	"/": {}, // Root directory
+	"":  {}, // Empty path
+}
+
+// semanticallyEqual checks if package and directory names are semantically equal to each other.
+func (PackageDirectoryMismatchRule) semanticallyEqual(packageName, dirName string) bool {
+	normDir := normalizePath(dirName)
+	normPkg := normalizePath(packageName)
+	return normDir == normPkg || normDir == "go"+normPkg
+}
```
Comment (ccoVeille) [inline rule/package_directory_mismatch.go:91]: There is something I dislike with the test here

If `od` package is in `good` folder it passes, no?

Comment (vbvictor) [reply rule/package_directory_mismatch.go:91]: It does.

Comment (vbvictor) [reply rule/package_directory_mismatch.go:91]: I think we don't want to directly enforce `go-foo` dir naming because it can be `go_foo` too. Maybe it's a reasonable trade-off to allow dir-`good` and package-`od`, will anyone even write like this?

Comment (vbvictor) [reply rule/package_directory_mismatch.go:91]: Looking at codesearch https://github.com/search?type=code&q=%22package+od+%22+language%3AGo nobody tried to name `package od`.

Comment (ccoVeille) [reply rule/package_directory_mismatch.go:91]: Of course,it would be weird.

What do you guys think we should do @alexandear @chavacava

File: RULES_DESCRIPTIONS.md [hunk 20]
```diff
@@ -963,6 +964,48 @@ More [information here](https://go.dev/wiki/CodeReviewComments#package-comments)
...
+
+Examples:
+
+Default behavior excludes paths containing "testdata"
+
+```toml
+[rule.package-directory-mismatch]
+```
+
+Exclude specific directories
```
Comment (chavacava) [inline RULES_DESCRIPTIONS.md]:
```suggestion
Ignore specific directories with `ignore-directories`
```

File: RULES_DESCRIPTIONS.md [hunk 21]
```diff
@@ -963,6 +964,48 @@ More [information here](https://go.dev/wiki/CodeReviewComments#package-comments)
...
+
+```toml
+[rule.package-directory-mismatch]
+arguments = [{ ignore-directories = ["testcases", "testinfo"] }]
+```
+
+Disable all exclusions
```
Comment (chavacava) [inline RULES_DESCRIPTIONS.md]:
```suggestion
Include all directories (`testdata` also)
```

File: RULES_DESCRIPTIONS.md [hunk 22]
```diff
@@ -963,6 +964,48 @@ More [information here](https://go.dev/wiki/CodeReviewComments#package-comments)
...
+Default behavior excludes paths containing "testdata"
+
+```toml
+[rule.package-directory-mismatch]
+```
+
+Exclude specific directories
+
+```toml
+[rule.package-directory-mismatch]
```
Comment (chavacava) [inline RULES_DESCRIPTIONS.md:998]: Running the rule with this configuration produces
```
initializing revive - getting lint rules: cannot configure rule: "package-directory-mismatch": invalid value [testcases testinfo] for argument ignore-directories of rule package-directory-mismatch, expected []string got []interface {}
```

File: RULES_DESCRIPTIONS.md [hunk 23]
```diff
@@ -963,6 +964,48 @@ More [information here](https://go.dev/wiki/CodeReviewComments#package-comments)
...
+arguments = [{ ignore-directories = ["testcases", "testinfo"] }]
+```
+
+Disable all exclusions
+
+```toml
+[rule.package-directory-mismatch]
```
Comment (chavacava) [inline RULES_DESCRIPTIONS.md:1005]: running the rule with this config produces
```
initializing revive - getting lint rules: cannot configure rule: "package-directory-mismatch": invalid value [] for argument ignoreDirectories of rule package-directory-mismatch, expected []string got []interface {}
```

File: rule/package_directory_mismatch.go [hunk 24]
```diff
@@ -0,0 +1,142 @@
...
+		if !ok {
+			return fmt.Errorf("invalid value %v for argument %s of rule %s, expected []string got %T", v, k, r.Name(), v)
+		}
+		return r.buildIgnoreRegex(ignoredDirs)
+	}
+
+	return nil
+}
+
+func (r *PackageDirectoryMismatchRule) buildIgnoreRegex(ignoredDirs []string) error {
```
Comment (chavacava) [inline rule/package_directory_mismatch.go]: This method name is misleading because the method _builds_ the regexp **and** _assigns_ it to `r.ignoredDirs`.

For the sake of clarity, I try to avoid methods with (hidden) side effects. I propose to change the signature to
```suggestion
func (*PackageDirectoryMismatchRule) buildIgnoreRegex(ignoredDirs []string) (*regexp.Regexp, error) {
```
and let the caller assign the result to `r.ignoredDirs`. Like that, all the updates of the rule config take place in a single method: `Configure`

File: rule/package_directory_mismatch.go (hunk 19 above)
Comment (chavacava) [reply rule/package_directory_mismatch.go:91]: IMO it's a corner case and we could live with it. 
I'm sure I do not well understand if the "normalization" stuff should be like it is... but I don't have another proposition (nor the motivation to fight with strings).

File: rule/package_directory_mismatch.go [hunk 25]
```diff
@@ -0,0 +1,142 @@
...
+	if len(ignoredDirs) == 0 {
+		return nil
+	}
+
+	patterns := make([]string, len(ignoredDirs))
+	for i, dir := range ignoredDirs {
+		patterns[i] = regexp.QuoteMeta(dir)
+	}
+	pattern := "(" + strings.Join(patterns, "|") + ")"
```
Comment (chavacava) [inline rule/package_directory_mismatch.go]: I think parenthesis are not needed here
```suggestion
	pattern := strings.Join(patterns, "|") 
```

File: rule/package_directory_mismatch.go [hunk 26]
```diff
@@ -0,0 +1,156 @@
...
+		return err
+	}
+
+	args, ok := arguments[0].(map[string]any)
+	if !ok {
+		return fmt.Errorf("invalid argument type: expected map[string]any, got %T", arguments[0])
+	}
+
+	for k, v := range args {
+		if !isRuleOption(k, "ignore-directories") {
```
Comment (alexandear) [inline rule/package_directory_mismatch.go]:
```suggestion
		if !isRuleOption(k, "ignoreDirectories") {
```

https://github.com/mgechev/revive/blob/b34c8e96eccd5e0cbcb03ff6a8b39419384b2a56/rule/enforce_switch_style.go#L31

https://github.com/mgechev/revive/blob/b34c8e96eccd5e0cbcb03ff6a8b39419384b2a56/rule/exported.go#L82

File: rule/package_directory_mismatch.go [hunk 27]
```diff
@@ -0,0 +1,156 @@
...
+	"":  {}, // Empty path
+}
+
+// semanticallyEqual checks if package and directory names are semantically equal to each other.
+func (PackageDirectoryMismatchRule) semanticallyEqual(packageName, dirName string) bool {
+	normDir := normalizePath(dirName)
+	normPkg := normalizePath(packageName)
+	return normDir == normPkg || normDir == "go"+normPkg
+}
+
```
Comment (ccoVeille) [inline rule/package_directory_mismatch.go:92]: > > In fact the call to normalizePath is not necessary (unless you do "go-")
> 
> It's still necessary to make this case work: `package foo_bar` with `dir go-foobar`. This testcase is missing, will add.

Can we add test cases with the following edge cases:

- go-foo-bar folder with foo_bar package
- go-foo-bar folder with foo-bar package
- go-foo-bar folder with foobar package
- go-foo_bar folder with foo_bar package
- go-foo_bar folder with foo-bar package
- go-foo_bar folder with foobar package
- go-od with good package
- go-od with of package
- good with go-od package

I feel like these marginal edges cases should all be tested for tested future regression.

We stated we agree they are edge cases, but I feel like we should find them back if we have a future refactoring

Comment (vbvictor) [reply rule/package_directory_mismatch.go:92]: Added these cases apart from `package foo-bar` which is invalid because packages can't have `-`.

Comment (ccoVeille) [reply rule/package_directory_mismatch.go:92]: Thanks, I appreciate

File: test/package_directory_mismatch_test.go [hunk 28]
```diff
@@ -0,0 +1,76 @@
...
+	testRule(t, "package_directory_mismatch/normalization/foo_bar/foobar", &rule.PackageDirectoryMismatchRule{}, config)
+	testRule(t, "package_directory_mismatch/normalization/foo.b_ar/foobar", &rule.PackageDirectoryMismatchRule{}, config)
+	testRule(t, "package_directory_mismatch/normalization/go-foo_bar/foo_bar", &rule.PackageDirectoryMismatchRule{}, config)
+	testRule(t, "package_directory_mismatch/normalization/go-foo_bar/foobar", &rule.PackageDirectoryMismatchRule{}, config)
+	testRule(t, "package_directory_mismatch/normalization/go-foo-bar/foo_bar", &rule.PackageDirectoryMismatchRule{}, config)
+	testRule(t, "package_directory_mismatch/normalization/go-foo-bar/foobar", &rule.PackageDirectoryMismatchRule{}, config)
+
+	testRule(t, "package_directory_mismatch/normalization/go-od/go_od", &rule.PackageDirectoryMismatchRule{}, config)
+	testRule(t, "package_directory_mismatch/normalization/go-od/good", &rule.PackageDirectoryMismatchRule{}, config)
+	testRule(t, "package_directory_mismatch/normalization/go-od/od", &rule.PackageDirectoryMismatchRule{}, config)
```
Comment (ccoVeille) [inline test/package_directory_mismatch_test.go:35]:
```suggestion
	// edge cases that are not logic, but we decided to ignore to have a simpler implementation 
	testRule(t, "package_directory_mismatch/normalization/go-od/go_od", &rule.PackageDirectoryMismatchRule{}, config)
	testRule(t, "package_directory_mismatch/normalization/go-od/good", &rule.PackageDirectoryMismatchRule{}, config)
	testRule(t, "package_directory_mismatch/normalization/go-od/od", &rule.PackageDirectoryMismatchRule{}, config)
```

**Review by chavacava — APPROVED** [review]

**Comment by chavacava** [comment]
Thank you @vbvictor for the PR and @alexandear @denisvmedia @ccoVeille for the feedback
